offload-items.lic

```diff
@@ -0,0 +1,159 @@
+    # Pause script and warn that the script is designed to offload items
+    # from your person to somewhere else. Item loss may occur.
+    DRC.message("WARNING: This script is designed to offload items from YOU to somewhere ELSE.")
+    DRC.message("WARNING: It also has the ability to move every item from the source container.")
+    DRC.message("WARNING: Item loss may occur. Use with caution.")
+    DRC.message("WARNING: If you want to transfer items between containers you own, use ;transfer-items instead.")
+    DRC.message("\n")
+    
+    unless noun
+      DRC.message("You are about to move EVERY SINGLE ITEM from from #{source} to #{destination}.")
+      DRC.message("Unpause the script to continue. Otherwise, ;kill the script to stop.")
+      pause_script
+      DRC.message("Are you SURE you want to offload EVERY SINGLE ITEM from YOUR #{source} to #{destination}?")
+      pause_script
+    else
+      DRC.message("You are about to move all #{noun}s from #{source} to #{destination}.")
+      DRC.message("Unpause the script to continue. Otherwise, ;kill the script to stop.")
+      pause_script
+    end
+    
+    # If container is very full then LOOK may not list all of them.
+    # If you're moving a specific item, then sort those to the top
+    # to increase chances we find and move all of them in one go.
+    bput("sort #{noun} in my #{source}", "are now at the top", "What were you referring to", "Please rephrase that command", "You may only sort items in your inventory") if noun
+    DRCI.get_item_list(source, 'look')
+      .map { |full_name| full_name =~ /lot of other stuff/ ? full_name : full_name.split(' ').last }
+      .select { |item| noun ? /\b#{noun}\b/ =~ item : true }
+      .each do |item|
+        # This indicates there is more items than LOOK can show
+        # and we've reached the end of what we last saw.
+        # We need to look again to see what we can now see.
+        # Keep doing this until fill the destination or exhaust source.
+        if item =~ /lot of other stuff/
+          offload_items(source, destination, noun)
+          return
+        end
+        # Attempt to get the item from the source container.
+        if DRCI.get_item(item, source)
+          if destination == 'trash'
+            trash_item(item)
+          else
+            unless move_item?(item, source, destination)
+              DRC.message("Unable to put #{item} in #{destination}. It may not accept items, or you may need to change the preposition (e.g. 'on shelf' vs. 'in shelf')")
+              # Failed. Return item to source container.
+              DRCI.put_away_item?(item, source)
+              exit
+            end
+          end
+        else
+          DRC.message("Unable to get #{item} from #{source}.")
+          DRC.message("Your hands are full!") if (DRC.left_hand && DRC.right_hand)
+          exit
+        end
+      end
+  end
+ 
+  def move_item?(item, source, destination)
```

Comment:
❓ Could the `move_item?` method be replaced by `DRCI.put_away_item_unsafe?(item, destination)`?

The `transfer-items` script uses `DRCI.put_away_item?` because it ensures the **my** prefix to the container to ensure you're transfering items on your person. In this case, you want to be able to put items in containers not on your person so no need to automatically add the **my** prefix, which is what the "unsafe" method variant allows.

Reply:
I considered this and actually initially designed the script around this helper method but the success patterns in DRCI don't properly consider the success messages we need when offloading items from our person to outside our person. It also doesn't properly handle putting items on|under|behind because it seems to be designed around putting items away on your person.

Also doesn't appear to properly handle my idea of including a preposition before the destination container such that you can specify "in shelf" or "on shelf" or "behind shelf" or "under shelf".

I considered modifying DRCI with helper methods to handle what I need, but figured that might be above my pay grade, so I handled it fully within this script.

Still, welcome suggestions on how I might utilize helper methods in DRC/DRCI.

Reply:
Gotcha. What you've proposed is fine to me, and we can iterate on if/how/when to propose changes into the DRC/DRCI helper methods to accommodate the new match strings in a future PR.

> Also doesn't appear to properly handle my idea of including a preposition before the destination container such that you can specify "in shelf" or "on shelf" or "behind shelf" or "under shelf".

I see what you mean now. The `DRCI.put_item_away?(..)` method hardcodes the `in` preposition. It needs to be refactored to be like `DRCI.get_item_unsafe(..)` that handles the in/on/under/etc.

> I considered modifying DRCI with helper methods to handle what I need, but figured that might be above my pay grade, so I handled it fully within this script.

Not at all! We have code reviews and tests to help folks new the codebase :)

You're thinking about all the right things 👍

Reply:
Let me clarify a little why I feel like `DRCI.put_away_item_unsafe?` does or doesn't do what I need and necessitates me writing my own method:

1. Hard-codes putting the item IN destination container (I think?). I want to support calling the script with "on shelf" and "in cabinet" and "behind desk", etc.
2. Related to above, doesn't allow us to understand if we've passed in a preposition
3. Doesn't include the proper @@success_patterns for offloading an item OFF your person, like `/^You set/i`

Reply:
The idea here is having a generic method that is used to limit the amount of methods you have to update down the line.  In that vein, my thoughts are adding an (optional) third  argument to the method 'preposition', which defaults to 'in':
- Should be OK as a third optional argument, as it would only make sense to include if you've also included a container.
- This should be non-breaking to anything already using it 
- Allows flexibility down the line where needed (like this)

```ruby
  def put_away_item_unsafe?(item, container = nil, preposition = 'in')
    command = "put #{item} #{preposition} #{container}" if container
    command = "stow #{item}" unless container
    result = DRC.bput(command, @@container_is_closed_patterns, @@put_away_item_success_patterns, @@put_away_item_failure_patterns, @@put_away_item_retry_patterns)
    case result
    when *@@container_is_closed_patterns
      return false unless open_container?(container)
      return put_away_item_unsafe?(item, container)
    when *@@put_away_item_retry_patterns
      return put_away_item_unsafe?(item, container)
    when *@@put_away_item_success_patterns
      return true
    when *@@put_away_item_failure_patterns
      return false
    else
      return false
    end
  end
```

Some additional testing error scenarios might need to be taken into account (wrong preposition, if you can't manipulate the container to open it).

Reply:
I like the idea of updating the DRCI method. The success patterns would need to be updated as well because DRCI currently doesn't account for successfully putting things elsewhere - e.g. "You set the cloak on the shelves". Do you have any concern intermingling success messaging for keeping items on your person (safe) with success messaging for offloading items off your person (unsafe)?

I'm just trying to be real thoughtful about item loss, is all.

Reply:
I don't think there should be any issues.  The patterns are just there to trap game output (and not end up with a 15 sec delay on command).  You've already executed the command itself, just trapping output.  

Keep in mind updates to common-* should be non-breaking to other scripts, and need to be rolled out in advance of regular scripts that depend on them, so you might need to break this into 2 PRs.

Reply:
Ok, you and Katoak have convinced me to do some work on DRCI. So my first PR was to combat-trainer of all things, and now my second PR will be to common-items. Wheee.

Reply:
> The idea here is having a generic method that is used to limit the amount of methods you have to update down the line if things from the game change. In that vein, my thoughts are adding an (optional) third argument to the method 'preposition', which defaults to 'in'

+1

Reply:
Handled via #5237